GH-38010: [Python] Construct pyarrow.Field and ChunkedArray through Arrow PyCapsule Protocol

jorisvandenbossche opened the pull request:
### Rationale for this change

See https://github.com/apache/arrow/issues/38010#issuecomment-2010601912 for more context. Right now for _consuming_ ArrowSchema-compatible objects that implement the PyCapsule interface, we only have the private `_import_from_c_capsule` (on Schema, Field, DataType) and we check for the protocol in the public `pa.schema(..)`.

But that means you currently can only consume objects that represent the schema of a batch (struct type), and not schemas of individual arrays. 

### What changes are included in this PR?

Expand the `pa.field(..)` constructor to accept objects implementing the protocol method.

### Are these changes tested?

TODO

* GitHub Issue: #38010

jorisvandenbossche commented:
cc @kylebarron @pitrou @wjones127

I don't particularly like the fact that doesn't really fit into the existing `pa.field(..)` signature, having to make other keywords optional / obsolete in case of passing a ArrowSchemaExportable object (also the `nullable` keyword would currently be silently ignored if you would specif that in this case) 

But the only alternative I can think of is adding something like a `pa.Field.from_arrow(..)` constructor method

pitrou commented:
I think this could be slightly smarter:
1. disallow `type`
2. if `nullable` or `metadata` is given, update the resulting field with the given metadata

jorisvandenbossche commented:
OK, updated this PR with the suggestion for the `pa.field()` API, and further also added support for the protocol in `pa.chunked_array()` (we recently added support to ChunkedArray for the C interface, but also here didn't yet have a public API to consume this).

With that, I think we have a public consume function or method for all cases (DataType is covered through `pa.field()`), and I also added a table to the documentation listing all the different functions that can be used to create pyarrow objects using the protocol.

pitrou commented on python/pyarrow/tests/test_cffi.py:
```diff
@@ -692,8 +692,14 @@ def test_roundtrip_chunked_array_capsule_requested_schema():
+
+    requested_type = pa.int64()
+    requested_capsule = requested_type.__arrow_c_schema__()
+    with pytest.raises(ValueError):
```
Check the error message as well?

pitrou commented on python/pyarrow/tests/test_table.py:
```diff
@@ -493,6 +493,47 @@ def test_recordbatch_dunder_init():
+    assert result == chunked
+
+    # Can also import with a type that implementer can cast to.
+    result = pa.chunked_array(wrapper, type=pa.int16())
```
Also test with a `TypeWrapper` that would wrap the `pa.int16` field and only expose `__arrow_c_schema__`?

pitrou commented on python/pyarrow/tests/test_types.py:
```diff
@@ -1335,3 +1335,25 @@ def __arrow_c_schema__(self):
+
+    assert pa.field(wrapped_field) == field
+
+    with pytest.raises(ValueError, match="cannot specify 'type'"):
```
Should be `TypeError`?

jorisvandenbossche commented on docs/source/python/extending_types.rst:
```diff
@@ -53,6 +53,31 @@ support for this protocol by checking for the presence of those methods, and
+can be used to create the various PyArrow objects:
+
++----------------------------+-----------------------------------------------+--------------------+
+| Result class               | Mapped Arrow type                             | Supported protocol |
```
Whoops, that's a left-over from the original table I copied from elsewhere for the table layout ;) 
Changed to "PyArrow constructor"

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1344,17 +1344,28 @@ cdef class ChunkedArray(_PandasConvertible):
+
+            if target_type != self.type:
+                try:
+                    chunked = self.cast(target_type, safe=True)
```
@paleolimbot added this on-the-fly casting for RecordBatchReader in https://github.com/apache/arrow/pull/40070, and it seems that for Table's stream export method, we convert this to a RecordBatchReader and that way Table also already supports casting chunk by chunk. 
We could indeed do something similar for ChunkedArray, will open an issue.

jorisvandenbossche replied:
Opened https://github.com/apache/arrow/issues/41064

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1437,6 +1451,21 @@ def chunked_array(arrays, type=None):
+            requested_type = None
+        capsule = arrays.__arrow_c_stream__(requested_type)
+        result = ChunkedArray._import_from_c_capsule(capsule)
+        if type is not None and result.type != type:
```
If `type` is None then `requested_type` is None, so it shouldn't matter which of the two I am checking here. But since this if clause is to check for whether we still need to cast to `type` (and not `requested_type`, which is a capsule not a pyarrow data type), it seems more logical to check for `type is not None`.

jorisvandenbossche commented:
@pitrou thanks for the review!

> Also test with a `TypeWrapper` that would wrap the `pa.int16` field and only expose `__arrow_c_schema__`?

In hindsight, we actually don't yet support this in any of our constructors (also `pa.array(..)` and `pa.table(..)` require an actual pyarrow type or schema). So reverted the test changes I did here and will leave this as a follow-up to tackle generally.
